Allow the user to specify the compression mode

flomnes opened the pull request:
Keep the old functions for compatibility.

Add new functions to allow for more flexibility : ZSTD seems pretty nice but can't be used with the current implementation.

ctabin commented on CMakeLists.txt:
```diff
@@ -59,7 +59,7 @@ export(TARGETS libzippp NAMESPACE libzippp:: FILE ${PROJECT_BINARY_DIR}/libzippp
 if(LIBZIPPP_BUILD_TESTS)
   enable_testing()
   add_executable(libzippp_test "tests/tests.cpp")
-  target_link_libraries(libzippp_test PRIVATE libzippp)
```
I don't get why this modification is related/needed for this PR ? :thinking:

flomnes replied:
I've removed it

ctabin commented on src/libzippp.h:
```diff
@@ -356,6 +356,7 @@ namespace libzippp {
          **/
         bool isEntryCompressionEnabled(const ZipEntry& entry) const;
         bool setEntryCompressionEnabled(const ZipEntry& entry, bool value) const;
```
This line has to be removed, the method doesn't exist anymore.

flomnes replied:
done

ctabin commented on tests/tests.cpp:
```diff
@@ -43,6 +43,7 @@
 #include <string>
 
 #include "libzippp.h"
+#include <zip.h> // ZIP_CM_DEFLATE
```
Actually the goal is to avoid import of `<zip.h>` all together. You should make some `#define` in libzippp.h that are mapped to the ones of libzip.

ctabin reviewed (changes requested):
Hi @flomnes, thanks for your PR. I left some comments.

ctabin commented on src/libzippp.h:
```diff
@@ -99,7 +99,26 @@ struct zip_source;
+   * Compression algorithm to use. See https://libzip.org/documentation/zip_set_file_compression.html
+   *
+   */
+  enum Compression {
```
```suggestion
  enum CompressionMethod {
```

ctabin commented on src/libzippp.h:
```diff
@@ -99,7 +99,26 @@ struct zip_source;
-    
+
+  /**
+   * Compression algorithm to use. See https://libzip.org/documentation/zip_set_file_compression.html
+   *
```
```suggestion
   * Compression algorithm to use.
   * See https://libzip.org/documentation/zip_set_file_compression.html
```

ctabin commented on src/libzippp.cpp:
```diff
@@ -47,6 +47,52 @@ using namespace std;
+
+static Compression convertCompressionFromLibzip(libzippp_uint16 comp) {
+  switch(comp) {
+  case ZIP_CM_STORE:
```
Since it is in the `switch`, please indent all the inner statements.

ctabin commented on src/libzippp.cpp:
```diff
@@ -47,6 +47,52 @@ using namespace std;
 
+static libzippp_uint16 convertCompressionToLibzip(Compression comp) {
+  switch(comp) {
+  case Compression::STORE:
```
Since it is in the `switch`, please indent all the inner statements.

ctabin reviewed (changes requested):
@flomnes This is great ! Juste a few remaining comments ad we're good.

Can you also please update the documentation in the `README` ?

flomnes commented:
Hi @ctabin, thank you for your feedback. I've applied your comments.

ctabin commented on src/libzippp.cpp:
```diff
@@ -385,17 +431,12 @@ bool ZipArchive::setComment(const string& comment) const {
-    
-    libzippp_uint16 compMode = value ? ZIP_CM_DEFLATE : ZIP_CM_STORE;
-    return zip_set_file_compression(zipHandle, entry.index, compMode, 0)==0;
+    const libzippp_uint16 comp_libzip = convertCompressionToLibzip(comp);
```
Looks like there is a `entry.compressionMethod = comp` missing here.

flomnes replied:
done

ctabin commented on README.md:
```diff
@@ -270,13 +274,19 @@ int main(int argc, char** argv) {
-
+  // Advanced usage : change the compression method. Default is DEFLATE.
+  if (useBzip2) {
+    auto entry = zf.getEntry("helloworld.txt");
```
```suggestion
    ZipEntry entry = zf.getEntry("helloworld.txt");
```

ctabin reviewed (changes requested):
@flomnes Thanks for the changes. Still some comments, but this looks very good !

ctabin commented:
@flomnes Is it possible for you to fix the conflicts ? It would be great to have this for the next release ! Thanks :pray:

flomnes commented:
Hi @ctabin,
- I fixed the merge conflicts
- I applied your remarks (had to remove some `const` qualifiers)
- I added a wrapper function `ZipArchive::setCompressionMethod`, that applies to all added entries. I find it easier to use than fetching the entry, setting its compression method.
- I added `const extern bool` flags to identify which compression methods are available. This may not be the best way, I'm open to suggestions here.

ctabin commented:
@flomnes That's amazing work ! Thanks for the PR. LGTM.
